core/src/test/java/org/apache/iceberg/TestPartitionSpecUpdate.java

```diff
@@ -0,0 +1,255 @@
+
+  @Test
+  public void testCommitUpdatedSpec() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addBucketField("id", 8)
+        .addBucketField("data", 16)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[0] + ": id_bucket: bucket[8](1)\n  " +
+        expectedFieldIds[1] + ": data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[2], table.spec().lastAssignedFieldId());
+
+    table.updateSpec().clear()
+        .addTruncateField("data", 8)
+        .commit();
+
+    Assert.assertEquals("[\n  " +
+        expectedFieldIds[3] + ": data_trunc: truncate[8](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(expectedFieldIds[4], table.spec().lastAssignedFieldId());
+  }
+
+  @Test
+  public void testUpdateException() {
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if there is an invalid partition field",
+        IllegalArgumentException.class, "Cannot use partition name more than once: id_bucket",
+        () -> table.updateSpec().clear()
+            .addBucketField("id", 8)
+            .addBucketField("id", 16)
+            .commit());
+  }
+
+  @Test
+  public void testAddDuplicateFieldException() {
+    AssertHelpers.assertThrows(
+        "Should throw IllegalArgumentException if adding a duplicate partition field",
+        IllegalArgumentException.class, "Cannot use partition name more than once: data_bucket",
+        () -> table.updateSpec()
+            .addBucketField("data", 16)
+            .commit());
+  }
+
+  @Test
+  public void testAddSamePartitionField() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+
+    if (formatVersion == 1) {
+      table.updateSpec()
+          .addBucketField("data", 16, "data_partition")
+          .commit();
+      Assert.assertEquals("[\n" +
+          "  1000: data_bucket: bucket[16](2)\n" +
+          "  1001: data_partition: bucket[16](2)\n" +
+          "]", table.spec().toString());
+    } else {
+      AssertHelpers.assertThrows(
+          "Should throw IllegalArgumentException if adding a duplicate partition field",
+          IllegalArgumentException.class,
+          "Field Id 1000 has already been used in the existing partition fields",
+          () -> table.updateSpec()
+              .addBucketField("data", 16, "data_partition")
+              .commit());
+    }
+  }
+
+  @Test
+  public void testAddField() {
+    Assert.assertEquals("[\n" +
+        "  1000: data_bucket: bucket[16](2)\n" +
+        "]", table.spec().toString());
+    Assert.assertEquals(1000, table.spec().lastAssignedFieldId());
+
+    table.updateSpec()
+        .addBucketField("data", 8, "data_partition")
```

Comment:
I think we should have a validation that rejects multiple bucket functions applied to the same table column. It doesn't make sense to do this.

I should note that there are some cases where you'd want to. Moving from `bucket[16]` to `bucket[32]`, for example, makes sense because the bucket function ensures that values are split between 2 buckets when moving from 16 to 32. So you could keep both in metadata without harm. But that isn't the case generally: if you had `bucket[7]` along with `bucket[11]`, you'd end up with 77 different combinations. If that's what the user wants, then they should use `bucket[77]`, not two separate bucket functions.

Reply:
@rdblue Thanks for the comments. One concern is that this change may break the existing table's partition spec if users already have created a table with a partition spec using multiple bucket functions for the same table column. 

We may just ignore existing violations and just enforce it for the new or updated specs by adding the check inside `bucket` method in `PartitionSpec.Builder`.